fix: Fix long parallel writings

dmga44 opened the pull request:
Versions of the test test_ec_parallel_writing with bigger file sizes
(65M) often fail. The reason behing those failures is the following
sequence of operations:
- some mountpoint starts writing some chunk X (some kilobyte of that
chunk), this triggers the creation of a chunk for that chunkIndex and
successfully finishes the write operation sending to the master an
update in the size of the file.
- other mountpoint (before the other one updates the file size) starts
writing some chunk Y, such that Y < X, i.e Y has a lower index than X.
Successfully finishes those writes, and at the moment of the updating
the file size, sends a size lower than current one by at least the
chunk X.
- in the master side, the update of the size (fsnodes_setlength
function) removes the chunks which are not necessary considering the
current file size, thus remving the chunk X and trashing the previous
writes on it.
- new writes on the X-th chunk allocate a new chunk and successfully
finishes the writing process, but the missing data is irreversible.

Solution is to extend the fsnodes_setlength to accept a
eraseFurtherChunks parameter to determine whether or not those chunks
should be erased. Summarizing:
- the client communication of write end
(SAU_CLTOMA_FUSE_WRITE_CHUNK_END header) will not remove chunks not
necessary given the size sent.
- the client communication of truncate (SAU_CLTOMA_FUSE_TRUNCATE
header) and truncate end (SAU_CLTOMA_FUSE_TRUNCATE_END header) will
continue removing chunks further than the given length.

As a side effect, the LENGTH metadata logs must be changed since the
state of the filesystem can not be recovered without saying whether
further chunks must be deleted or not:
- old metadata will be handled as always removing further chunks.
- new metadata will have 0|1 to represent whether the further chunks
must be deleted.

A test extending this issue was added in the LongSystemTests suite.

BREAKING CHANGE: changes the changelog format generated by master
server.

An initial tool for downgrading the system from the v5.0.0 to v4.X.X was
uploaded with all the previous changes.

dmga44 committed:
fix!: Fix long parallel writings

Versions of the test test_ec_parallel_writing with bigger file sizes
(65M) often fail. The reason behing those failures is the following
sequence of operations:
- some mountpoint starts writing some chunk X (some kilobyte of that
chunk), this triggers the creation of a chunk for that chunkIndex and
successfully finishes the write operation sending to the master an
update in the size of the file.
- other mountpoint (before the other one updates the file size) starts
writing some chunk Y, such that Y < X, i.e Y has a lower index than X.
Successfully finishes those writes, and at the moment of the updating
the file size, sends a size lower than current one by at least the
chunk X.
- in the master side, the update of the size (fsnodes_setlength
function) removes the chunks which are not necessary considering the
current file size, thus remving the chunk X and trashing the previous
writes on it.
- new writes on the X-th chunk allocate a new chunk and successfully
finishes the writing process, but the missing data is irreversible.

Solution is to extend the fsnodes_setlength to accept a
eraseFurtherChunks parameter to determine whether or not those chunks
should be erased. Summarizing:
- the client communication of write end
(SAU_CLTOMA_FUSE_WRITE_CHUNK_END header) will not remove chunks not
necessary given the size sent.
- the client communication of truncate (SAU_CLTOMA_FUSE_TRUNCATE
header) and truncate end (SAU_CLTOMA_FUSE_TRUNCATE_END header) will
continue removing chunks further than the given length.

As a side effect, the LENGTH metadata logs must be changed since the
state of the filesystem can not be recovered without saying whether
further chunks must be deleted or not:
- old metadata will be handled as always removing further chunks.
- new metadata will have 0|1 to represent whether the further chunks
must be deleted.

BREAKING CHANGE: changes the changelog format generated by master
server.

Signed-off-by: Dave <dave@leil.io>

dmga44 committed:
test: Test long parallel writings

This test is a version of the test_ec_parallel_writing with file size
of 513MB. A test template for both tests were created to simplify
them.

Signed-off-by: Dave <dave@leil.io>

rolysr commented on tests/test_suites/ShortSystemTests/test_ec_parallel_writing.sh:
```diff
@@ -1,42 +1 @@
-saunafs_chunkserver_daemon $csid stop
-
-MESSAGE="Parity is corrupted after writing" expect_success file-validate "$file"
+FILE_SIZE_MB=25 source test_suites/TestTemplates/test_ec_parallel_writing.inc
```
Why was 25 the value for FILE_SIZE_MB?

dmga44 replied:
Honestly, don't know for sure. It was the previous size of the file used. My guesses are the following considering some very small block size, such as 1KB and making the writes in parallel and random, then the time to finish those operations is quite near the time the test should last.

rolysr reviewed (dismissed):
LGTM. Just a minor doubt

dmga44 commented on tests/test_suites/TestTemplates/test_ec_parallel_writing.inc:
```diff
@@ -0,0 +1,41 @@
+saunafs setgoal ec_3_1 "$file"
+
+# Create a temporary file with 1 gigabyte plus 1 megabyte of generated data
+tmpf=$RAMDISK_DIR/tmpf
```
Applied.

dmga44 commented on tests/test_suites/TestTemplates/test_ec_parallel_writing.inc:
```diff
@@ -0,0 +1,41 @@
+	(
+		file="${info[mount${i}]}/file"
+		seq $i 10 $((${FILE_SIZE_MB}*1024-1)) | shuf | expect_success xargs -P5 -IXX \
+				dd if="$tmpf" of="$file" bs=1K count=1 seek=XX skip=XX conv=notrunc 2>/dev/null
```
Applied.

dmga44 commented on tests/test_suites/TestTemplates/test_ec_parallel_writing.inc:
```diff
@@ -0,0 +1,41 @@
+wait
+
+# Validate the result
+MESSAGE="Data is corrupted after writing" expect_success file-validate "$file"
```
Applied.

dmga44 commented on tests/test_suites/TestTemplates/test_ec_parallel_writing.inc:
```diff
@@ -0,0 +1,41 @@
+
+# Create a temporary file with 1 gigabyte plus 1 megabyte of generated data
+tmpf=$RAMDISK_DIR/tmpf
+FILE_SIZE=${FILE_SIZE_MB}M file-generate "$tmpf"
```
Applied.

uristdwarf commented:
Can you mark the commits as breaking change, e.g:

`fix!: Fix long parallel writings`

And add at the end a BREAKING CHANGE footer.

See here for an example

https://www.conventionalcommits.org/en/v1.0.0/

ralcolea commented on src/master/filesystem_operations.cc:
```diff
@@ -641,8 +641,11 @@ uint8_t fs_do_setlength(const FsContext &context, uint32_t inode, uint64_t lengt
 
-	fsnodes_setlength(static_cast<FSNodeFile*>(p), length);
-	fs_changelog(ts, "LENGTH(%" PRIu32 ",%" PRIu64 ")", inode, static_cast<FSNodeFile*>(p)->length);
+	bool eraseFurtherChunks = true;
```
Consider to explain the reason of `eraseFurtherChunks = true`.
I think it's not trivial, specially because the reason does not rely in `fs_do_setlength()` function itself, but in the function which call `fs_do_setlength()`, like `matoclserv_chunk_status()` and `matoclserv_fuse_truncate`.

dmga44 replied:
Applied.

ralcolea commented on src/master/filesystem_operations.cc:
```diff
@@ -2019,11 +2023,14 @@ uint8_t fs_writeend(uint32_t inode, uint64_t length, uint64_t chunkid, uint32_t
 		}
 		if (length > p->length) {
-			fsnodes_setlength(p, length);
+			bool eraseFurtherChunks = false;
```
Same suggestion here about explaining the reason of `eraseFurtherChunks = false`.

dmga44 replied:
Applied.

ralcolea reviewed (dismissed):
Great job @dmga44 :+1: :fire: :rocket:

dmga44 committed:
chore: Add migrations structure

This change targets providing a place to store everything related to
migrations, which is specially important if there are breaking changes
involved. The idea is to fill the summary of the breaking changes and
the scripts for downgrading a specific service with the upcoming
breaking changes. The structure is also helpful for the next versions
upgrade scheme.

Changes:
- added a folder in the top level for the migration tools.
- added a folder inside migrations for the upgrade from v4.X.X to
v5.0.0, and another one inside of it for downgrade tools.
- added Breaking-changes-summary.md.
- added downgrade-master.sh which is the only service that needs a
special downgrade behavior for now.

Signed-off-by: Dave <dave@leil.io>

wg1k commented on tests/test_suites/TestTemplates/test_ec_parallel_writing.inc:
```diff
@@ -0,0 +1,41 @@
+if is_windows_system; then
+	timeout_set $(((${FILE_SIZE_MB} / 25) * 5)) minutes
```
When using arithmetic expansion, you can refer to variables without the usual dollar sign. For example, you could write it like this:
```suggestion
	timeout_set $(( (FILE_SIZE_MB / 25) * 5 )) minutes
```
Both styles work fine, so feel free to use whichever one.

dmga44 replied:
Applied.

wg1k commented on migrations/v4.X.X-to-v5.0.0/rollback/downgrade-master.sh:
```diff
@@ -0,0 +1,145 @@
+	# Start the master server
+	log_message "Starting master server..."
+	systemctl start saunafs-master
+	if [ $? -eq 0 ]; then
```
Check exit code directly with e.g. 'if mycmd;', not indirectly with $?.

dmga44 replied:
Applied.

wg1k commented on migrations/v4.X.X-to-v5.0.0/rollback/downgrade-master.sh:
```diff
@@ -0,0 +1,145 @@
+	# Downgrade the master server to the specified version
+	log_message "Downgrading master server to version ${TARGET_VERSION}..."
+	apt install saunafs-master=${TARGET_VERSION}
+	if [ $? -eq 0 ]; then
```
Check exit code directly with e.g. 'if mycmd;', not indirectly with $?.

dmga44 replied:
Applied.

wg1k commented on migrations/v4.X.X-to-v5.0.0/rollback/downgrade-master.sh:
```diff
@@ -0,0 +1,145 @@
+stop_master_server() {
+	# Stop the master server
+	systemctl stop saunafs-master
+	if [ $? -eq 0 ]; then
```
Check exit code directly with e.g. 'if mycmd;', not indirectly with $?.

dmga44 replied:
Applied.

wg1k commented on migrations/v4.X.X-to-v5.0.0/rollback/downgrade-master.sh:
```diff
@@ -0,0 +1,145 @@
+		\?)
+			echo "Invalid option: -${OPTARG}" >&2
+			usage
+			exit 1
```
As the usage is already exiting, this line seems unreachable.

dmga44 replied:
Applied.

wg1k commented on migrations/v4.X.X-to-v5.0.0/rollback/downgrade-master.sh:
```diff
@@ -0,0 +1,145 @@
+		:)
+			echo "Option -${OPTARG} requires an argument." >&2
+			usage
+			exit 1
```
As the usage is already exiting, this line seems unreachable.

dmga44 replied:
Applied.

wg1k commented on migrations/v4.X.X-to-v5.0.0/rollback/downgrade-master.sh:
```diff
@@ -0,0 +1,145 @@
+	# Apply fixes to the installation
+	log_message "Applying fixes to the installation..."
+	# Fix the LENGTH function in changelog.sfs files
+	for file in $(find ${MASTER_DATA_DIR} -type f -name "changelog.sfs*"); do
```
For loops over find output are fragile. Use find -exec or a while read loop:
```shell
	find "${MASTER_DATA_DIR}" -type f -name "changelog.sfs*" | \
		while IFS= read -r file; do
```

dmga44 replied:
Applied.

wg1k commented on migrations/v4.X.X-to-v5.0.0/rollback/downgrade-master.sh:
```diff
@@ -0,0 +1,145 @@
+MASTER_DATA_DIR=/var/lib/saunafs
+
+while getopts "v:b:d:" opt; do
+	case $opt in
```
Prefer putting braces around variable references even when not strictly required.
`$opt` -> `${opt}`

dmga44 replied:
Applied.

wg1k reviewed (approved):
:+1:
